src/dataflow/src/source/timestamp.rs

```diff
@@ -165,33 +280,71 @@ impl TimestampBindingBox {
         }
     }
 
-    fn add_binding(&mut self, partition: PartitionId, timestamp: Timestamp, offset: MzOffset) {
-        let partition = self
-            .partitions
-            .entry(partition)
-            .or_insert_with(PartitionTimestamps::new);
-        partition.add_binding(timestamp, offset);
+    fn add_partition(&mut self, partition: PartitionId) {
+        if self.partitions.contains_key(&partition) {
+            debug!("already inserted partition {:?}, ignoring", partition);
+            return;
+        }
+
+        self.partitions
+            .insert(partition, PartitionTimestamps::new());
+    }
+
+    fn add_binding(
+        &mut self,
+        partition: PartitionId,
+        timestamp: Timestamp,
+        offset: MzOffset,
+        proposed: bool,
+    ) {
+        if !self.partitions.contains_key(&partition) {
+            error!("missing partition {:?} when adding binding", partition);
+            return;
+        }
+
+        if proposed {
+            if let Some(proposer) = &mut self.proposer {
+                proposer.propose_binding(partition, timestamp, offset);
+            } else {
+                error!(
+                    "attempting to propose a timestamp binding on a source that isn't real-time."
+                );
+            }
+        } else {
+            let partition = self.partitions.get_mut(&partition).expect("known to exist");
+            partition.add_binding(timestamp, offset);
+        }
     }
 
     fn get_binding(
         &self,
         partition: &PartitionId,
         offset: MzOffset,
-    ) -> Option<(Timestamp, MzOffset)> {
+    ) -> Option<(Timestamp, Option<MzOffset>)> {
         if !self.partitions.contains_key(partition) {
             return None;
         }
 
         let partition = self.partitions.get(partition).expect("known to exist");
-        partition.get_binding(offset)
+        if let Some((time, offset)) = partition.get_binding(offset) {
+            Some((time, Some(offset)))
+        } else if let Some(proposer) = &self.proposer {
+            Some((proposer.upper(), None))
+        } else {
+            None
+        }
     }
 
     fn read_upper(&self, target: &mut Antichain<Timestamp>) {
         target.clear();
```

Comment:
this API intentionally mirrors the differential API here https://docs.rs/differential-dataflow/0.12.0/differential_dataflow/trace/trait.TraceReader.html#method.read_upper and I suspect its like that to avoid an allocation

Reply:
Interesting, thanks!